src/zenml/login/credentials.py

```diff
@@ -0,0 +1,342 @@
+
+from pydantic import BaseModel, ConfigDict
+
+from zenml.login.pro.constants import ZENML_PRO_API_URL, ZENML_PRO_URL
+from zenml.login.pro.tenant.models import TenantRead, TenantStatus
+from zenml.models import ServerModel
+from zenml.services.service_status import ServiceState
+from zenml.utils.enum_utils import StrEnum
+from zenml.utils.string_utils import get_human_readable_time
+
+
+class ServerType(StrEnum):
+    """The type of server."""
+
+    PRO_API = "PRO_API"
+    PRO = "PRO"
+    REMOTE = "REMOTE"
+    LOCAL = "LOCAL"
+
+
+class APIToken(BaseModel):
+    """Cached API Token."""
+
+    access_token: str
+    expires_in: Optional[int] = None
+    expires_at: Optional[datetime] = None
+    leeway: Optional[int] = None
+    cookie_name: Optional[str] = None
+    device_id: Optional[UUID] = None
+    device_metadata: Optional[Dict[str, Any]] = None
+
+    @property
+    def expires_at_with_leeway(self) -> Optional[datetime]:
+        """Get the token expiration time with leeway.
+
+        Returns:
+            The token expiration time with leeway.
+        """
+        if not self.expires_at:
+            return None
+        if not self.leeway:
+            return self.expires_at
+        return self.expires_at - timedelta(seconds=self.leeway)
+
+    @property
+    def expired(self) -> bool:
+        """Check if the token is expired.
+
+        Returns:
+            bool: True if the token is expired, False otherwise.
+        """
+        expires_at = self.expires_at_with_leeway
+        if not expires_at:
+            return False
+        return expires_at < datetime.now(timezone.utc)
+
+    model_config = ConfigDict(
+        # Allow extra attributes to allow backwards compatibility
+        extra="allow",
+    )
+
+
+class ServerCredentials(BaseModel):
+    """Cached Server Credentials."""
+
+    url: str
+    api_key: Optional[str] = None
+    api_token: Optional[APIToken] = None
+    username: Optional[str] = None
+    password: Optional[str] = None
+
+    # Extra server attributes
+    server_id: Optional[UUID] = None
+    server_name: Optional[str] = None
+    organization_name: Optional[str] = None
+    organization_id: Optional[UUID] = None
+    status: Optional[str] = None
+    version: Optional[str] = None
+
+    @property
+    def id(self) -> str:
+        """Get the server identifier.
+
+        Returns:
+            The server identifier.
+        """
+        if self.server_id:
+            return str(self.server_id)
+        return self.url
+
+    @property
+    def type(self) -> ServerType:
+        """Get the server type.
+
+        Returns:
+            The server type.
+        """
+        from zenml.login.pro.utils import is_zenml_pro_server_url
+
+        if self.url == ZENML_PRO_API_URL:
+            return ServerType.PRO_API
+        if self.organization_id or is_zenml_pro_server_url(self.url):
+            return ServerType.PRO
+        if urlparse(self.url).hostname in ["localhost", "127.0.0.1"]:
```

Comment:
This will still return `ServerType.REMOTE` for `host.docker.internal` URLs. I'm not sure if that matters at all, just seems like `LOCAL` would be correct?

Reply:
Correct